src/StaticLS/HIE/File.hs

```diff
@@ -0,0 +1,173 @@
+module StaticLS.HIE.File where
+
+import Control.Exception
+import Control.Monad
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.IO.Unlift
+import Control.Monad.Trans (lift)
+import Control.Monad.Trans.Maybe (MaybeT (..), runMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask)
+import Data.Either (fromRight)
+import Data.List
+import Data.List.Extra (dropPrefix)
+import qualified Data.Map as Map
+import Data.Maybe (catMaybes, mapMaybe)
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import qualified GHC
+import GHC.Data.Maybe (firstJusts)
+import qualified GHC.Iface.Ext.Binary as GHC
+import qualified GHC.Iface.Ext.Types as GHC
+import qualified GHC.Types.Name.Cache as GHC
+import qualified GHC.Unit.Types as GHC
+import HieDb
+import qualified Language.LSP.Types as LSP
+import StaticLS.StaticEnv
+import System.Directory (doesFileExist, makeAbsolute)
+import System.FilePath (normalise, (-<.>), (</>))
+
+type SrcFilePath = FilePath
+type HieFilePath = FilePath
+
+-- | Retrieve a hie info from a lsp text document identifier
+getHieFileFromTdi :: (HasStaticEnv m, MonadIO m) => LSP.TextDocumentIdentifier -> m (Maybe GHC.HieFile)
+getHieFileFromTdi tdi = do
+    staticEnv <- getStaticEnv
+    runMaybeT $ do
+        srcFilePath <- MaybeT $ pure $ LSP.uriToFilePath tdi._uri
+        hieFilePath <- MaybeT $ srcFilePathToHieFilePath srcFilePath
+        MaybeT $ getHieFile hieFilePath
+
+getHieFile :: (HasStaticEnv m, MonadIO m) => HieFilePath -> m (Maybe GHC.HieFile)
+getHieFile hieFilePath = do
+    staticEnv <- getStaticEnv
+    liftIO $
+        (Just <$> (pure . GHC.hie_file_result <=< GHC.readHieFile staticEnv.nameCache) hieFilePath)
+            -- Return nothing if the file failed to read
+            `catch` (\e -> let _ = (e :: IOException) in pure Nothing)
+
+modToHieFile :: (HasStaticEnv m, MonadIO m) => GHC.ModuleName -> m (Maybe GHC.HieFile)
+modToHieFile = runMaybeT . (MaybeT . getHieFile <=< MaybeT . modToHieFilePath)
+
+modToSrcFile :: (HasStaticEnv m, MonadIO m) => GHC.ModuleName -> m (Maybe SrcFilePath)
+modToSrcFile = runMaybeT . (MaybeT . hieFilePathToSrcFilePath <=< MaybeT . modToHieFilePath)
+
+{- | Fetch a src file from an hie file, checking hiedb but falling back on a file manipulation method
+if not indexed
+-}
+srcFilePathToHieFilePath :: (HasStaticEnv m, MonadIO m) => SrcFilePath -> m (Maybe HieFilePath)
+srcFilePathToHieFilePath srcPath = do
+    t1 <- srcFilePathToHieFilePathHieDb srcPath
+    t2 <- srcFilePathToHieFilePathFromFile srcPath
+    pure $ firstJusts [t1, t2]
+
+hieFilePathToSrcFilePath :: (HasStaticEnv m, MonadIO m) => HieFilePath -> m (Maybe SrcFilePath)
+hieFilePathToSrcFilePath = hieFilePathToSrcFilePathFromFile
```

Comment:
what is the motivation for this synonym?

Reply:
Ideally I would like to be able to actually fetch the src path from hiedb but no utility function exists for doing that yet. Instead right now I directly read the hie file which is quite a bit slower on large sets. Using the synonym wherever this is called lets me substitute the implementation out more easily or do something similar to the src to hie file where I use multiple fallback.